timeutil/schedule.go

```diff
@@ -82,40 +91,267 @@ func ParseTime(s string) (t TimeOfDay, err error) {
+	}
+}
+
+// Subspans returns a slice of TimeSpan generated from ts by splitting the time
+// between ts.Start and ts.End into ts.Split equal spans.
+func (ts ClockSpan) Subspans() []ClockSpan {
+	if ts.Split == 0 || ts.Split == 1 || ts.End == ts.Start {
+		return []ClockSpan{ts}
+	}
+
+	span := ts.End.Sub(ts.Start)
+	step := span / time.Duration(ts.Split)
+
+	spans := make([]ClockSpan, ts.Split)
+	for i := uint(0); i < ts.Split; i++ {
+		start := ts.Start.Add(time.Duration(i) * step)
+		spans[i] = ClockSpan{
+			Start:  start,
+			End:    start.Add(step),
+			Split:  0, // no more subspans
+			Spread: ts.Spread,
+		}
+	}
+	return spans
+}
 
-	Weekday string
+// Schedule represents a single schedule
+type Schedule struct {
+	WeekSpans  []WeekSpan
+	ClockSpans []ClockSpan
 }
 
 func (sched *Schedule) String() string {
-	return fmt.Sprintf("%s-%s", sched.Start, sched.End)
+	buf := &bytes.Buffer{}
+
+	for i, span := range sched.WeekSpans {
+		if i > 0 {
+			buf.WriteByte(',')
+		}
+		buf.WriteString(span.String())
+	}
+
+	if len(sched.WeekSpans) > 0 && len(sched.ClockSpans) > 0 {
+		buf.WriteByte(',')
+	}
+
+	for i, span := range sched.ClockSpans {
+		if i > 0 {
+			buf.WriteByte(',')
+		}
+		buf.WriteString(span.String())
+	}
+	return buf.String()
 }
 
-func (sched *Schedule) Next(last time.Time) (start, end time.Time) {
-	now := timeNow()
+func (sched *Schedule) flattenedTimeSpans() []ClockSpan {
+	baseTimes := sched.ClockSpans
+	if len(baseTimes) == 0 {
+		baseTimes = []ClockSpan{{}}
+	}
 
-	t := last
-	for {
-		a := time.Date(t.Year(), t.Month(), t.Day(), sched.Start.Hour, sched.Start.Minute, 0, 0, time.Local)
-		b := time.Date(t.Year(), t.Month(), t.Day(), sched.End.Hour, sched.End.Minute, 0, 0, time.Local)
+	times := make([]ClockSpan, 0, len(baseTimes))
+	for _, ts := range baseTimes {
+		times = append(times, ts.Subspans()...)
+	}
+	return times
+}
 
-		// not using AddDate() here as this can panic() if no
-		// location is set
-		t = t.Add(24 * time.Hour)
+// isLastWeekdayInMonth returns true if t.Weekday() is the last weekday
+// occurring this t.Month(), eg. check is Feb 25 2017 is the last Saturday of
+// February.
+func isLastWeekdayInMonth(t time.Time) bool {
+	// try a week from now, if it's still the same month then t.Weekday() is
+	// not last
+	return t.Month() != t.Add(7*24*time.Hour).Month()
+}
 
-		// same inteval as last update, move forward
-		if (last.Equal(a) || last.After(a)) && (last.Equal(b) || last.Before(b)) {
-			continue
+// ScheduleWindow represents a time window between Start and End times when the
+// scheduled event can happen.
+type ScheduleWindow struct {
+	Start time.Time
+	End   time.Time
+	// Spread defines whether the event shall be randomly placed between
+	// Start and End times
+	Spread bool
+}
+
+// Includes returns whether t falls inside the window.
+func (s ScheduleWindow) Includes(t time.Time) bool {
+	return (t.Equal(s.Start) || t.After(s.Start)) && (t.Equal(s.End) || t.Before(s.End))
```

Comment:
more compact - check if we're outside of the start-end range: `return !(t.Before(s.Start) || t.After(s.End))`